File: Problems/Strings/Valid-Parentheses/Valid-Parentheses.cs
using System;
using System.Collections.Generic;

// Using a stack - Time O(n)
public class Solution
{
    public bool IsValid(string s)
    {
        Stack<char> stack = new Stack<char>();
        foreach (char c in s)
        {
            if (c == '(' || c == '{' || c == '[')
            {
                stack.Push(c);
            }
            else if (stack.Count == 0 || (c == ')' && stack.Peek() != '(') || (c == '}' && stack.Peek() != '{')
                || (c == ']' && stack.Peek() != '['))
            {
                return false;
            }
            else
            {
                stack.Pop();
            }
        }
        return stack.Count == 0;
    }
}

class Program
{
    static void Main(string[] args)
    {
        string input = "({[()]})"; 
        Console.WriteLine("Input: s =" + input);

        Solution sol = new Solution();
        bool isValid = sol.IsValid(input);

        Console.WriteLine("Output: " + isValid);
    }
}

File: Problems/Strings/Valid-Parentheses/Valid-Parentheses.cpp
#include <iostream>
#include <stack>
#include <string>

// Using a stack - Time O(n)

class Solution {
public:
    bool isValid(std::string s)
    {
        std::stack<char> stack;
        for (char c : s)
        {
            if (c == '(' || c == '{' || c == '[')
                stack.push(c);
            else if (stack.empty() || (c == ')' && stack.top() != '(') || (c == '}' && stack.top() != '{') 
                || (c == ']' && stack.top() != '['))
                return false;
            else
                stack.pop();
        }
        return stack.empty();
    }
};

int main()
{
    std::string input = "({[()]})"; 
    std::cout << "Input: s =" << input << std::endl;

    Solution sol;
    bool isValid = sol.isValid(input);

    std::cout << "Output: " << std::boolalpha << isValid  << std::endl;

    return 0;
}
